AFQ/api.py

```diff
@@ -71,135 +72,216 @@ def do_preprocessing():
                             and templates.get(name + hemi + '_prob_map')):
-                        afq_bundles[name + hemi] = {
-                            'ROIs': [templates[name + '_roi1' + hemi],
-                                     templates[name + '_roi2' + hemi]],
+                        bundle = {
+                            'ROIs': [
+                                templates[name + '_roi1' + hemi],
+                                templates[name + '_roi2' + hemi]],
                             'rules': [True, True],
-                            'prob_map': templates[name + hemi + '_prob_map'],
+                            'prob_map': templates[
+                                name + hemi + '_prob_map'],
                             'cross_midline': False,
-                            'uid': uid}
+                            'uid': self.uid_dict[key]}
                     else:
-                        logger.warning(f"{name} is not in AFQ templates")
-
-                    uid += 1
-    elif seg_algo.startswith("reco"):
-        if seg_algo.endswith("80"):
-            bundle_dict = afd.read_hcp_atlas(80)
-        else:
-            bundle_dict = afd.read_hcp_atlas(16)
-
-        afq_bundles = {}
-        uid = 1
-        afq_bundles["whole_brain"] = bundle_dict["whole_brain"]
-        for name in bundle_names:
-            if name in RECO_UNIQUE:
-                afq_bundles[name] = bundle_dict[name]
-                afq_bundles[name]['uid'] = uid
-                uid += 1
+                        raise ValueError(f"{key} is not in AFQ templates")
+                if self.resample_to:
+                    for ii, roi in enumerate(bundle['ROIs']):
+                        bundle['ROIs'][ii] =\
+                            afd.read_resample_roi(
+                                roi, resample_to=self.resample_to)
+                self._dict[key] = bundle
+        elif self.seg_algo.startswith("reco"):
+            if self.seg_algo.endswith("80"):
+                reco_bundle_dict = afd.read_hcp_atlas(80)
             else:
-                for hemi in ["_R", "_L"]:
-                    afq_bundles[name + hemi] = bundle_dict[name + hemi]
-                    afq_bundles[name + hemi]['uid'] = uid
-                    uid += 1
-    else:
-        raise ValueError("Input: %s is not a valid input`seg_algo`" % seg_algo)
+                reco_bundle_dict = afd.read_hcp_atlas(16)
+            for key in self.bundle_names:
+                bundle = reco_bundle_dict[key]
+                bundle['uid'] = self.uid_dict[key]
+                self._dict[key] = bundle
+        else:
+            raise ValueError(
+                "Input: %s is not a valid input`seg_algo`" % self.seg_algo)
+        self.all_gen = True
+
+    def __setitem__(self, key, item):
+        self._dict[key] = item
+        self.uid_dict[key] = self.c_uid
+        self.c_uid += 1
+        self.bundle_names.append(key)
```

Comment:
Do you need to reset `all_gen` to `False` here, if it's currently set to `True`?

Reply:
I clarified the use of __setitem__